src/JSInterop/Microsoft.JSInterop/src/DotNetDispatcher.cs

```diff
@@ -127,30 +127,38 @@ public static void BeginInvoke(string callId, string assemblyName, string method
             }
         }
 
-        private static object InvokeSynchronously(string assemblyName, string methodIdentifier, object targetInstance, string argsJson)
+        private static object InvokeSynchronously(string assemblyName, string methodIdentifier, IDotNetObjectRef objectReference, string argsJson)
         {
             AssemblyKey assemblyKey;
-            if (targetInstance != null)
+            if (objectReference is null)
+            {
+                assemblyKey = new AssemblyKey(assemblyName);
+            }
+            else
             {
                 if (assemblyName != null)
                 {
                     throw new ArgumentException($"For instance method calls, '{nameof(assemblyName)}' should be null. Value received: '{assemblyName}'.");
                 }
 
-                assemblyKey = new AssemblyKey(targetInstance.GetType().Assembly);
-            }
-            else
-            {
-                assemblyKey = new AssemblyKey(assemblyName);
+                if (string.Equals("__Dispose", methodIdentifier, StringComparison.Ordinal))
+                {
+                    // The client executed dotNetObjectReference.dispose(). Dispose the reference and exit.
+                    objectReference.Dispose();
+                    return default;
+                }
+
+                assemblyKey = new AssemblyKey(objectReference.Value.GetType().Assembly);
             }
 
+
             var (methodInfo, parameterTypes) = GetCachedMethodInfo(assemblyKey, methodIdentifier);
 
             var suppliedArgs = ParseArguments(methodIdentifier, argsJson, parameterTypes);
 
             try
             {
-                return methodInfo.Invoke(targetInstance, suppliedArgs);
+                return methodInfo.Invoke(objectReference?.Value, suppliedArgs);
```

Comment:
It's probably worth a comment here that the `objectReference?.Value` is expected to be null in the non-instance-method case. I had to read this a few times to convince myself it's right.